File: identhub/src/main/java/de/solarisbank/identhub/verfication/bank/VerificationBankFragment.java
package de.solarisbank.identhub.verfication.bank;

import android.graphics.drawable.LevelListDrawable;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import de.solarisbank.identhub.ViewModelFactory;
import de.solarisbank.identhub.R;
import de.solarisbank.identhub.databinding.FragmentVerificationBankBinding;
import de.solarisbank.identhub.di.LibraryComponent;
import de.solarisbank.identhub.BaseFragment;
import de.solarisbank.identhub.identity.IdentityActivityViewModel;

import static de.solarisbank.identhub.verfication.bank.VerificationBankViewModel.DISABLED_STATE;
import static de.solarisbank.identhub.verfication.bank.VerificationBankViewModel.ERROR_STATE;

public class VerificationBankFragment extends BaseFragment {

    ViewModelFactory viewModelFactory;

    private VerificationBankViewModel viewModel;

    private IdentityActivityViewModel sharedViewModel;

    private FragmentVerificationBankBinding binding;

    public static Fragment newInstance() {
        return new VerificationBankFragment();
    }

    @Override
    protected void inject(LibraryComponent component) {
        component.inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentVerificationBankBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initViewModel();
        initViews();
        observeInputsState();
    }

    private void initViewModel() {
        viewModel = new ViewModelProvider(this, viewModelFactory)
                .get(VerificationBankViewModel.class);

        sharedViewModel = new ViewModelProvider(requireActivity(), viewModelFactory)
                .get(IdentityActivityViewModel.class);
    }

    private void initViews() {
        binding.submitButton.setOnClickListener(view -> viewModel.onSubmitButtonClicked());
    }

    private void observeInputsState() {
        viewModel.getState().observe(getViewLifecycleOwner(), this::onStateInputChanged);
    }

    private void onStateInputChanged(Integer value) {
        int state = value;
        ((LevelListDrawable) binding.ibanNumber.getBackground()).setLevel(state);
        binding.errorMessage.setVisibility(state == ERROR_STATE ? View.VISIBLE : View.GONE);

        if (value == DISABLED_STATE) {
            sharedViewModel.navigateTo(R.id.action_verificationBankFragment_to_establishConnectionFragment);
        }
    }
}

File: identhub/src/main/java/de/solarisbank/identhub/verfication/bank/VerificationBankFragment.kt
package de.solarisbank.identhub.verfication.bank

import android.graphics.drawable.LevelListDrawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import com.jakewharton.rxbinding2.view.RxView
import com.jakewharton.rxbinding2.widget.RxTextView
import com.jakewharton.rxbinding2.widget.TextViewTextChangeEvent
import de.solarisbank.identhub.base.IdentHubFragment
import de.solarisbank.identhub.databinding.FragmentVerificationBankBinding
import de.solarisbank.identhub.di.FragmentComponent
import de.solarisbank.identhub.identity.IdentityActivityViewModel
import de.solarisbank.identhub.verfication.bank.VerificationBankViewModel.IBanState
import de.solarisbank.sdk.core.activityViewModels
import de.solarisbank.sdk.core.result.*
import de.solarisbank.sdk.core.result.Type.ResourceNotFound
import de.solarisbank.sdk.core.view.viewBinding
import de.solarisbank.sdk.core.viewModels
import io.reactivex.disposables.CompositeDisposable
import timber.log.Timber

class VerificationBankFragment : IdentHubFragment() {
    private val compositeDisposable: CompositeDisposable = CompositeDisposable()
    private val binding: FragmentVerificationBankBinding by viewBinding { FragmentVerificationBankBinding.inflate(layoutInflater) }
    private val sharedViewModel: IdentityActivityViewModel by lazy<IdentityActivityViewModel> { activityViewModels() }
    private val viewModel: VerificationBankViewModel by lazy<VerificationBankViewModel> { viewModels() }

    override fun inject(component: FragmentComponent) {
        component.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViews()
        observeInputsState()
        observeVerifyResult()
    }

    private fun observeVerifyResult() {
        viewModel.getVerifyResultLiveData().observe(viewLifecycleOwner, Observer { onResultVerifyIBanChanged(it) })
    }

    private fun onResultVerifyIBanChanged(result: Result<String>) {
        if (result.succeeded) {
            val resultData = result.data
            sharedViewModel.moveToEstablishSecureConnection(resultData)
        } else if (result is Result.Error) {
            val type = result.typ
            if (type is Type.ServerError || type is ResourceNotFound) {
                sharedViewModel.navigateToVerificationBankError()
            } else {
                updateIBanInputState(IBanState.INVALID)
            }
        }
    }

    private fun updateIBanInputState(iBanState: IBanState) {
        (binding.ibanNumber.background as LevelListDrawable).level = iBanState.value
        binding.errorMessage.visibility = if (iBanState === IBanState.INVALID) View.VISIBLE else View.GONE
    }

    private fun initViews() {
        compositeDisposable.add(RxView.clicks(binding.submitButton)
                .map { binding.ibanNumber.text.toString().trim { it <= ' ' } }
                .filter { viewModel.validationIBan(it) }
                .subscribe(
                        { iBan: String -> viewModel.onSubmitButtonClicked(iBan) },
                        { throwable: Throwable? -> Timber.e(throwable, "Cannot valid IBAN") })
        )
        compositeDisposable.add(RxTextView.textChangeEvents(binding.ibanNumber)
                .subscribe { textViewTextChangeEvent: TextViewTextChangeEvent -> viewModel.onIBanInputChanged(textViewTextChangeEvent.text().toString()) })
    }

    private fun observeInputsState() {
        viewModel.iBanState.observe(viewLifecycleOwner, Observer { event: Event<IBanState> -> onIBanInputValidationStateChanged(event) })
    }

    private fun onIBanInputValidationStateChanged(event: Event<IBanState>) {
        val iBanState = event.content
        iBanState?.let { updateIBanInputState(it) }
    }

    override fun onDestroyView() {
        compositeDisposable.clear()
        super.onDestroyView()
    }

    companion object {
        fun newInstance(): Fragment {
            return VerificationBankFragment()
        }
    }
}